### PR TITLE
3911 gaps immunisation program back end query

### LINKED ISSUE #3911: GAPS: vaccine course back end: query
Part of epic #3662 

A vaccine course briefly known as a Immunisation Program defines the relationship between a demographic and what items and schedules are due to be completed. Multiple vaccine courses can be defined within one program.

## Acceptance criteria
* [x] Table created
* [x] Service layer, diesel layer ( no filter, no sorting )
* [x] graphQL query to return list of programs

## Describe the solution you'd like 🎁

Create tables for the immunisation program and rows (IP)
Create the diesel and service layers for querying data. Default sort is by IP name, no sort or filter required. Return all columns in the service layer
Create graphQL query

```mermaid
erDiagram
    vaccine_course {
        TEXT id PK "NOT NULL"
        TEXT immunisation_program FK "REFERENCES program(id)"
        TEXT name "NOT NULL"
        TEXT demographic_indicator_id FK "NOT NULL"
        FLOAT coverage_rate "NOT NULL DEFAULT 100"
        BOOL is_active "NOT NULL DEFAULT true"
        FLOAT wastage_rate "NOT NULL DEFAULT 0"
        INT doses ""
    }
    vaccine_course_items {
        TEXT vaccine_course_id FK "NOT NULL REFERENCES vaccine_course(id)"
        TEXT item_link_id FK "NOT NULL REFERENCES item_link(id)"
    }
    vaccine_course_schedule {
        TEXT id PK "NOT NULL"
        TEXT vaccine_course_id FK "NOT NULL REFERENCES vaccine_course(id)"
        INT dose_number "NOT NULL"
        INT label "E.g. day 0, day 42, etc"
    }
    program {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"  
        TEXT master_list_id 
    }
vaccine_course ||--o{ program : ""
vaccine_course_items ||--o{ vaccine_course : ""
vaccine_course_schedule ||--o{ vaccine_course : ""
```

### Additional context 💌

| User story  | B.2.1 |
| :------------- | :------------- |
As a… | National EPI manager
I want… | to be able to detail the key requirements of each individual immunization program
So that… |the cold storage demand for each immunization program can be tracked independently based on their unique set of requirements

**Solution B.2.1**

The system will provide a mechanism for central users to define an unlimited number of customisable immunization programs. 
These immunization programs will form the framework for determining estimated cold storage capacity demand. The contents of the framework are detailed in the following User Stories. 

| User story  | B.2.1.1 |
| :------------- | :------------- |
As a… | National EPI manager
I want… | to be able to select which immunization programs should be included in demand calculations
So that… | I can easily run hypothetical gap analysis and compare what happens if I include a program or not in the calculations

**Solution B.2.1.1**

The system will provide a mechanism for central users to include, or not include, immunization programs when estimating cold storage capacity demand. 

| User story  | B.2.3 |
| :------------- | :------------- |
As a… | National EPI manager
I want… | to be able to specify which demographic group each immunization program targets
So that… | our forecasting takes into account the number of people that require immunizing for each immunization program 

**Solution B.2.3**

The system will allow central users to associate demographic groups with immunization programs. 
Only 1 demographic group can be linked with each immunization program. 

| User story  | B.2.4 |
| :------------- | :------------- |
As a… | National EPI manager
I want… | to be able to specify one or more vaccines which are required for each immunization program
So that… | our forecasting takes into account the type of vaccines involved for each immunization program

**Solution B.2.4**

The system will allow central users to associate one or multiple vaccines with each immunization program. 
The vaccines will be available to choose from the Master Data product catalogue. 
The same vaccine may be associated with more than one immunization program. 

| User story  | B.2.5 |
| :------------- | :------------- |
As a… | National EPI manager
I want… | to be able to specify how many doses of each vaccine is required to fully immunize a person in the target demographic
So that… | our forecasting takes into account the total number of doses required for each immunization program

**Solution B.2.5**

The system will allow central users to specify the number of doses required to immunize the demographic group associated with each immunization program. 
This data is required to calculate the total cold storage capacity required per person per immunization program.

| User story  | B.2.6 |
| :------------- | :------------- |
As a… | National EPI manager
I want… | to be able to specify a 'coverage' percentage rate for each immunization program
So that… | I can easily apply a modifier to each immunization program to reflect real-world adjustments

**Solution B.2.6**

The system will allow central users to specify a ‘coverage’ rate for each immunization program. 
The figure will default to 100% but may be modified by the central user as required to model real-world factors. 

### Moneyworks Jobcode 🧰
`CCEIDEV`

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #3911 

# 👩🏻‍💻 What does this PR do?

Creates `vaccine_course`, `vaccine_course_item`, and `vaccine_course_schedule` tables

Basic service layer and graphql.

## 💌 Any notes for the reviewer?

I'm thinking the graphql can be extended to handle the needs of the frontend, so kept it basic for now.

Are there any Validations I need to add?

# 🧪 Testing

Backend only will be tested as part of the epic

# 📃 Documentation

- [X] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

**Review by lache-melvin — APPROVED** [review]
Looks great thanks James! A few nits, and I think worth adding validation steps for program, and demographics indicator (maybe have to do this one later though?) but working great 🔥

File: server/graphql/vaccine_course/src/types/vaccine_course.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    pub async fn program_id(&self) -> &str {
+        &self.row().program_id
+    }
+
+    pub async fn demographic_indicator_id(&self) -> &str {
+        &self.row().demographic_indicator_id
+    }
+    // TODO Loaders for the program and demographic_indicator
```
Comment (fergie-nz) [inline server/graphql/vaccine_course/src/types/vaccine_course.rs:35]: I wonder am I understanding right that you would also have loaders for vaccine course schedule and items tables?

File: server/repository/src/db_diesel/vaccine_course/vaccine_course_row.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+)]
+#[diesel(table_name = vaccine_course)]
+pub struct VaccineCourseRow {
+    pub id: String,
+    pub name: String,
+    pub program_id: String,
+    pub demographic_indicator_id: String,
+    pub coverage_rate: f64,
+    pub is_active: bool,
+    pub wastage_rate: f64,
```
Comment (fergie-nz) [inline server/repository/src/db_diesel/vaccine_course/vaccine_course_row.rs:35]: I noticed we don't have wastage rate on the front end. I wonder if we will use this elsewhere?
Maybe generic to a facility /  or facility level so we could apply different rates of wastage for the same course at different levels.

Maybe you had come to a conclusion about this elsewhere?

File: server/repository/src/migrations/v2_01_00/activity_log.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+use crate::migrations::DATETIME;
+use crate::{migrations::sql, StorageConnection};
+
+pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
+    if cfg!(feature = "postgres") {
+        sql!(
+            connection,
+            r#"
+                ALTER TYPE activity_log_type ADD VALUE 'ASSET_PROPERTY_CREATED';
```
Comment (fergie-nz) [inline server/repository/src/migrations/v2_01_00/activity_log.rs:9]: Is this right? Or just missing from the asset properties PR?

**Review by fergie-nz — APPROVED** [review]
Nice James 🚒 

Only thing I am left wondering about is the wastage rate and if it should be in these tables

File: server/service/src/vaccine_course/insert.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+) -> VaccineCourseRow {
+    VaccineCourseRow {
+        id,
+        name,
+        program_id,
+        demographic_indicator_id,
+        coverage_rate: 100.0,
+        is_active: true,
+        wastage_rate: 0.0,
+        doses: 1,
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply server/service/src/vaccine_course/insert.rs:80]: I think these are logical defaults. My plan is to allow them to be updated in an update, but we'll keep creation simple. We won't be re-using the same modal for update.

File: server/repository/src/migrations/v2_01_00/activity_log.rs (hunk 3 above)
Comment (jmbrunskill) [reply server/repository/src/migrations/v2_01_00/activity_log.rs:9]: Missing from the other PR I think. Should have commented

File: server/repository/src/db_diesel/vaccine_course/vaccine_course_row.rs (hunk 2 above)
Comment (adamdewey) [reply server/repository/src/db_diesel/vaccine_course/vaccine_course_row.rs:35]: Hey @fergie-nz - the wastage rates needs to be applied at the `vaccine course` level - we'll need a way for central users to add a percentage value via the UI

File: server/graphql/vaccine_course/src/types/vaccine_course.rs (hunk 1 above)
Comment (jmbrunskill) [reply server/graphql/vaccine_course/src/types/vaccine_course.rs:35]: Yes you're right. I think we will.

File: server/service/src/auth.rs [hunk 5]
```diff
@@ -524,6 +526,14 @@ fn all_permissions() -> HashMap<Resource, PermissionDSL> {
         Resource::QueryAsset,
         PermissionDSL::HasPermission(PermissionType::AssetQuery),
     );
+    map.insert(
+        Resource::MutateVaccineCourse,
+        PermissionDSL::HasPermission(PermissionType::VaccineCourseMutate),
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply server/service/src/auth.rs:531]: I'll create a new issue for Permissions...

**Pull request merged**